ystia/yorc
# Feature/brbdcf 1043 hosts pool from file

# Pull Request description

## Description of the change

Added a command to apply a Hosts Pool configuration provided in a YAML or JSON file :
```bash
yorc hostspool apply [--auto-approve] <filename>
```
And a command to export the Hosts Pool configuration as a YAML or JSON representation, to the standard output or a file:
```bash
yorc hostspool export [--output yaml|json] [--file <filename>]
```

The apply command  accepts YAML or JSON files. The following properties are supported :

  * ``hosts``: List of hosts configuration. A host configuration supports the following properties,
     - ``name``: mandatory string identifying the host, no other host entry can have the same name value in the file
     - ``connection``: Connection configuration,
        + ``host``: Hostname or ip address used to connect to the host (defaults to the ``name`` described above)
        + ``user``: name of the user used to connect to the host (default "root")
        + ``password``: either a password or a private key should be provided
        + ``private_key``: Path to a private key file (or private key file content), either a password or a private key should be provided
        + ``port``: Port used to connect to the host (default 22)
     - ``labels``: key/value pairs

Example of a YAML Hosts Pool configuration file :

```YAML

    hosts:
    - name: host1
      connection:
        host: host1.example.com
        user: test
        private_key: /path/to/secrets/id_rsa
        port: 22
      labels:
        environment: dev
        testlabel: hello
        host.cpu_frequency: 3 GHz
        host.disk_size: 50 GB
        host.mem_size: 4GB
        host.num_cpus: "4"
        os.architecture: x86_64
        os.distribution: ubuntu
        os.type: linux
        os.version: "17.1"
    - name: host2
      connection:
        host: host2.example.com
        user: test
        password: test
```

The apply will compare and display the differences between the current Hosts Pool configuration and the configuration specified in the file. A user confirmation will be asked before proceeding, except if --auto-approve option is specified.

### How I did it

In prov/hostspool/hostspool_mgr.go, added a new function Apply(pool []Host, checkpoint *uint64)
This function will compare the current Hosts Pool configuration with the Hosts Pool in argument and apply changes. The checkpoint in argument, if not nil, will be used to verify the changes are applied to the expected Hosts Pool configuration. The checkpoint value for the current Hosts Pool configuration is returned by the function List() implemented in the same file.
Internally, this checkpoint corresponds to a consul ModifyIndex. 
As a consequence the lock key kvLocckKey initially created under the hosts_pool subtree had to be moved outside of this subtree, or the command apply by just locking this lock was changing the ModifyIndex, making the checkpoint in argument invalid.

Factorized the code already existing for CRUD operations on one host, to have common code computing necessary  operations that the new Apply function will apply in a single transaction.

### How to verify it

Preliminary step: environment setup

To test a Hosts Pool in a development environment, you can use a dockerized ssh server.
First as a yorc user, generate private/public keys in a <path to secrets>/ directory, check the private key is read-only for your yorc user,
then run the following commands to have 3 test hosts host1, host2, host3 that can be added to a Hosts Pool :
(--privileged option below is needed because the container has also docker-in-docker capability, although it is not used here)
```bash
docker run --privileged -p 8701:22 \
 -e "AUTH_KEY=$(cat <path to secrets>/id_rsa.pub)" \
 --rm -d --name host1 --hostname host1 laurentg/testhost

docker run --privileged -p 8702:22 \
 -e "AUTH_KEY=$(cat <path to secrets>/id_rsa.pub)" \
 --rm -d --name host2 --hostname host2 laurentg/testhost

docker run --privileged -p 8703:22 \
 -e "AUTH_KEY=$(cat <path to secrets>/id_rsa.pub)" \
 --rm -d --name host3 --hostname host3 laurentg/testhost
```

Verify that as yorc user, you can connect to each host using the private key you generated :
```bash
ssh -i <path to secrets>/id_rsa test@myhost -p 8701
ssh -i <path to secrets>/id_rsa test@myhost -p 8702
ssh -i <path to secrets>/id_rsa test@myhost-p 8703
```

Hosts Pool apply tests

You can get examples of yaml/json files already created and update connection settings to use your host:

Configuration with 1 host : https://github.com/laurentganne/ystia-devenv/blob/master/testlab/pool1host.yaml
Configuration with an additional host and label changes for the existing host : https://github.com/laurentganne/ystia-devenv/blob/master/testlab/pool2hosts.yaml
Same configuration with label diffs, in json : https://github.com/laurentganne/ystia-devenv/blob/master/testlab/pool2hosts.json
Configuration with an additional host : https://github.com/laurentganne/ystia-devenv/blob/master/testlab/pool3hosts.yaml

First apply pool1host.yaml :

```bash
yorc hp apply pool1host.yaml
```

Check you get a confirmation message showing host1 will be added.
Enter y

Check you have 1 host now in the pool, running :

```bash
yorc hp list
```

Then apply pool2host.yaml

```bash
yorc hp apply pool2host.yaml
```

Check the confirmation message shows host2 will be added, and 2 host1 labels will be updated and 1 host1 label will be removed
Enter y

Then apply pool2host.json

```bash
yorc hp apply pool2host.json
```

Check there will be label changes on host1 and host2

Then apply pool3host.yaml but don't answer to the confirmation question.
Just stay in this interactive step.

```bash
yorc hp apply pool3host.yaml
```

Instead of accepting the change, launch another window, and form this new window, launch a command to used the first configuration with 1 host :


```bash
yorc hp apply pool1host.yaml
```

Check you are warned that host2 will be removed. Enter y.

Go back to the first window where you attempted to add a third host to the pool that had 2 hosts at that time.
Enter y
Check you get an error message that detected a Hosts Pool checkpoint issue, as the Hosts Pool configuration has changed since, and the operation is refused.

Try again the command :

```bash
yorc hp apply pool3host.yaml
```

Check the confirmation message correctly states that host2 and host3 will be added.
Enter n (or nothing) to refuse the change.

Stop the host2 running this command :

```bash
docker stop host2
```

Try to add host2 to the pool :

```bash
yorc hp apply pool2host.yaml
```
Enter y

Check you get a message that the apply was done successfully but that a connection failure was detected for host2.

Start again host2 :

```bash
docker run --privileged -p 8702:22 \
 -e "AUTH_KEY=$(cat <path to secrets>/id_rsa.pub)" \
 --rm -d --name host2 --hostname host2 laurentg/testhost
```

Verify that as yorc user, you can connect to host2: 
```bash
ssh -i <path to secrets>/id_rsa test@myhost -p 8702
```

Apply the json hosts pool configuration that just updated labels :

```bash
yorc hp apply pool2hosts.json
```

Run yorc hp list and check host2 is back online with no connection error

Export checks :

Run this command to see the Hosts Pool configuration exported in YAML representation :

```bash
yorc hp export
```

Run this command to have a json output :

```bash
yorc hp export -o json
```

Run this command to save the Hosts Pool configuration in a yaml file :

```bash
yorc hp export -f tmp.yaml
```
